runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/sideinput/SideInputStateFetcher.java

```diff
@@ -0,0 +1,215 @@
+   * was cached.
+   *
+   * <p>Returns {@literal null} if the side input was not ready, {@literal Optional.absent()} if the
+   * side input was null, and {@literal Optional.present(...)} if the side input was non-null.
+   */
+  @SuppressWarnings("deprecation")
+  public <T> SideInput<T> fetchSideInput(
+      PCollectionView<T> view,
+      BoundedWindow sideWindow,
+      String stateFamily,
+      SideInputState state,
+      Supplier<Closeable> scopedReadStateSupplier) {
+    Callable<SideInput<T>> loadSideInputFromWindmill =
+        () -> loadSideInputFromWindmill(view, sideWindow, stateFamily, scopedReadStateSupplier);
+
+    SideInputCache.Key sideInputCacheKey =
+        SideInputCache.Key.create(view.getTagInternal(), sideWindow);
+
+    try {
+      if (state == SideInputState.KNOWN_READY) {
+        Optional<SideInput<T>> existingCacheEntry = sideInputCache.get(sideInputCacheKey);
+        if (!existingCacheEntry.isPresent()) {
+          return sideInputCache.getOrLoad(sideInputCacheKey, loadSideInputFromWindmill);
+        }
+
+        if (!existingCacheEntry.get().isReady()) {
+          return sideInputCache.invalidateThenLoadNewEntry(
+              sideInputCacheKey, loadSideInputFromWindmill);
+        }
+
+        return existingCacheEntry.get();
+      }
+
+      return sideInputCache.getOrLoad(sideInputCacheKey, loadSideInputFromWindmill);
+    } catch (Exception e) {
+      LOG.error("Fetch failed: ", e);
+      throw new RuntimeException("Exception while fetching side input: ", e);
+    }
+  }
+
+  @SuppressWarnings({"deprecation", "unchecked"})
+  private <T, SideWindowT extends BoundedWindow> GlobalData fetchGlobalDataFromWindmill(
+      PCollectionView<T> view,
+      SideWindowT sideWindow,
+      String stateFamily,
+      Supplier<Closeable> scopedReadStateSupplier)
+      throws IOException {
+    WindowingStrategy<?, SideWindowT> sideWindowStrategy =
+        (WindowingStrategy<?, SideWindowT>) view.getWindowingStrategyInternal();
+
+    Coder<SideWindowT> windowCoder = sideWindowStrategy.getWindowFn().windowCoder();
+
+    ByteStringOutputStream windowStream = new ByteStringOutputStream();
+    windowCoder.encode(sideWindow, windowStream, Coder.Context.OUTER);
+
+    Windmill.GlobalDataRequest request =
+        Windmill.GlobalDataRequest.newBuilder()
+            .setDataId(
+                Windmill.GlobalDataId.newBuilder()
+                    .setTag(view.getTagInternal().getId())
+                    .setVersion(windowStream.toByteString())
+                    .build())
+            .setStateFamily(stateFamily)
+            .setExistenceWatermarkDeadline(
+                WindmillTimeUtils.harnessToWindmillTimestamp(
+                    sideWindowStrategy.getTrigger().getWatermarkThatGuaranteesFiring(sideWindow)))
+            .build();
+
+    try (Closeable ignored = scopedReadStateSupplier.get()) {
+      return server.getSideInputData(request);
+    }
+  }
+
+  @SuppressWarnings("deprecation")
```

Comment:
localize and comment (for all suppressions)

Reply:
done!

a lot of the SuppressWarnings around deprecation is needed for the side input cache, and creating the values/keys for the cache.
if there is a non-deprecated way of getting that data, i can change it to the new way and remove the annotation.